File: YouTubePlayerView.swift
//
//  YouTubePlayerView.swift
//  YouTubePlayer
//
//  A self-contained UIView subclass that plays YouTube videos in-app
//  using WKWebView and the YouTube IFrame Player API.
//  No redirects to YouTube app or Safari.
//
//  Swift 5 | iOS 12.0+
//

import UIKit
import WebKit

// MARK: - Player State Enum

/// Represents the current state of the YouTube player
public enum YouTubePlayerState: Int {
    case unstarted = -1
    case ended = 0
    case playing = 1
    case paused = 2
    case buffering = 3
    case cued = 5
}

// MARK: - Player Quality Enum

/// Video quality options for YouTube playback
public enum YouTubePlaybackQuality: String {
    case small = "small"
    case medium = "medium"
    case large = "large"
    case hd720 = "hd720"
    case hd1080 = "hd1080"
    case highRes = "highres"
    case auto = "default"
}

// MARK: - Delegate Protocol

/// Delegate protocol for receiving player events
public protocol YouTubePlayerViewDelegate: AnyObject {
    func playerReady(_ playerView: YouTubePlayerView)
    func player(_ playerView: YouTubePlayerView, didChangeStateTo state: YouTubePlayerState)
    func player(_ playerView: YouTubePlayerView, didChangeQualityTo quality: YouTubePlaybackQuality)
    func player(_ playerView: YouTubePlayerView, didReceiveError error: YouTubePlayerError)
    func player(_ playerView: YouTubePlayerView, didPlayTime time: Float)
}

// Default implementations (all optional)
public extension YouTubePlayerViewDelegate {
    func playerReady(_ playerView: YouTubePlayerView) {}
    func player(_ playerView: YouTubePlayerView, didChangeStateTo state: YouTubePlayerState) {}
    func player(_ playerView: YouTubePlayerView, didChangeQualityTo quality: YouTubePlaybackQuality) {}
    func player(_ playerView: YouTubePlayerView, didReceiveError error: YouTubePlayerError) {}
    func player(_ playerView: YouTubePlayerView, didPlayTime time: Float) {}
}

// MARK: - Player Error Enum

/// Errors that can occur during YouTube playback
public enum YouTubePlayerError: Int {
    case invalidParameter = 2
    case html5Error = 5
    case videoNotFound = 100
    case notEmbeddable = 101
    case notEmbeddableAlt = 150
    case unknown = -1
}

// MARK: - YouTubePlayerView

/// A UIView subclass that embeds a YouTube player using WKWebView and IFrame API.
/// All playback occurs within the app - no redirects to YouTube app or Safari.
public class YouTubePlayerView: UIView {

    // MARK: - Properties

    /// Delegate for receiving player events
    public weak var delegate: YouTubePlayerViewDelegate?

    /// The WKWebView used for playback
    private var webView: WKWebView!

    /// Current video ID being played
    private(set) public var currentVideoId: String?

    /// Whether the player is ready to receive commands
    private(set) public var isReady: Bool = false

    /// Configuration for the player
    private var playerVars: [String: Any] = [:]

    // MARK: - Initialization

    public override init(frame: CGRect) {
        super.init(frame: frame)
        setupWebView()
    }

    public required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupWebView()
    }

    /// Initialize with custom player variables
    /// - Parameter playerVars: Dictionary of YouTube IFrame player parameters
    public convenience init(frame: CGRect, playerVars: [String: Any]) {
        self.init(frame: frame)
        self.playerVars = playerVars
    }

    // MARK: - Setup

    private func setupWebView() {
        // Configure WKWebView to allow inline playback and autoplay
        let config = WKWebViewConfiguration()
        config.allowsInlineMediaPlayback = true
        config.mediaTypesRequiringUserActionForPlayback = []

        // Create web view
        webView = WKWebView(frame: bounds, configuration: config)
        webView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        webView.scrollView.isScrollEnabled = false
        webView.scrollView.bounces = false
        webView.isOpaque = false
        webView.backgroundColor = .black

        // Set navigation delegate to intercept external links
        webView.navigationDelegate = self

        // Add message handler for JavaScript callbacks
        webView.configuration.userContentController.add(LeakAvoider(delegate: self), name: "youtubePlayer")

        addSubview(webView)
    }

    // MARK: - Public Methods

    /// Load and prepare a YouTube video for playback
    /// - Parameter videoId: The YouTube video ID (e.g., "dQw4w9WgXcQ")
    public func load(videoId: String) {
        currentVideoId = videoId
        isReady = false

        let html = generateHTML(videoId: videoId)
        webView.loadHTMLString(html, baseURL: URL(string: "https://www.youtube.com"))
    }

    /// Load a video by full YouTube URL
    /// - Parameter url: Full YouTube URL (supports youtube.com/watch and youtu.be formats)
    public func load(url: String) {
        if let videoId = extractVideoId(from: url) {
            load(videoId: videoId)
        }
    }

    /// Start or resume video playback
    public func play() {
        evaluatePlayerCommand("player.playVideo();")
    }

    /// Pause video playback
    public func pause() {
        evaluatePlayerCommand("player.pauseVideo();")
    }

    /// Stop video playback
    public func stop() {
        evaluatePlayerCommand("player.stopVideo();")
    }

    /// Seek to a specific time in the video
    /// - Parameters:
    ///   - seconds: Time to seek to in seconds
    ///   - allowSeekAhead: Whether to allow seeking beyond buffered content
    public func seek(to seconds: Float, allowSeekAhead: Bool = true) {
        evaluatePlayerCommand("player.seekTo(\(seconds), \(allowSeekAhead));")
    }

    /// Mute the video
    public func mute() {
        evaluatePlayerCommand("player.mute();")
    }

    /// Unmute the video
    public func unmute() {
        evaluatePlayerCommand("player.unMute();")
    }

    /// Set playback volume
    /// - Parameter volume: Volume level (0-100)
    public func setVolume(_ volume: Int) {
        let clampedVolume = max(0, min(100, volume))
        evaluatePlayerCommand("player.setVolume(\(clampedVolume));")
    }

    /// Set playback rate
    /// - Parameter rate: Playback rate (0.25, 0.5, 1, 1.5, 2)
    public func setPlaybackRate(_ rate: Float) {
        evaluatePlayerCommand("player.setPlaybackRate(\(rate));")
    }

    /// Set video quality
    /// - Parameter quality: Desired playback quality
    public func setPlaybackQuality(_ quality: YouTubePlaybackQuality) {
        evaluatePlayerCommand("player.setPlaybackQuality('\(quality.rawValue)');")
    }

    /// Get current playback time
    /// - Parameter completion: Callback with current time in seconds
    public func getCurrentTime(completion: @escaping (Float) -> Void) {
        webView.evaluateJavaScript("player.getCurrentTime();") { result, _ in
            if let time = result as? NSNumber {
                completion(time.floatValue)
            }
        }
    }

    /// Get video duration
    /// - Parameter completion: Callback with duration in seconds
    public func getDuration(completion: @escaping (Float) -> Void) {
        webView.evaluateJavaScript("player.getDuration();") { result, _ in
            if let duration = result as? NSNumber {
                completion(duration.floatValue)
            }
        }
    }

    /// Load a new video and start playing
    /// - Parameter videoId: The YouTube video ID
    public func cueVideo(videoId: String) {
        currentVideoId = videoId
        evaluatePlayerCommand("player.cueVideoById('\(videoId)');")
    }

    /// Load a new video and start playing immediately
    /// - Parameter videoId: The YouTube video ID
    public func loadVideo(videoId: String) {
        currentVideoId = videoId
        evaluatePlayerCommand("player.loadVideoById('\(videoId)');")
    }

    // MARK: - Fullscreen Support

    /// Present the player in fullscreen mode
    /// - Parameter viewController: The view controller to present from
    public func enterFullscreen(from viewController: UIViewController) {
        let fullscreenVC = YouTubeFullscreenViewController(playerView: self)
        fullscreenVC.modalPresentationStyle = .fullScreen
        viewController.present(fullscreenVC, animated: true)
    }

    // MARK: - Private Methods

    private func evaluatePlayerCommand(_ command: String) {
        guard isReady else { return }
        webView.evaluateJavaScript(command, completionHandler: nil)
    }

    private func extractVideoId(from url: String) -> String? {
        // Handle youtu.be short URLs
        if url.contains("youtu.be/") {
            return url.components(separatedBy: "youtu.be/").last?.components(separatedBy: "?").first
        }

        // Handle youtube.com/watch URLs
        if let urlComponents = URLComponents(string: url),
           let queryItems = urlComponents.queryItems,
           let videoId = queryItems.first(where: { $0.name == "v" })?.value {
            return videoId
        }

        // Handle youtube.com/embed URLs
        if url.contains("/embed/") {
            return url.components(separatedBy: "/embed/").last?.components(separatedBy: "?").first
        }

        return nil
    }

    private func generateHTML(videoId: String) -> String {
        // Merge default player vars with custom ones
        var vars: [String: Any] = [
            "playsinline": 1,
            "controls": 1,
            "showinfo": 0,
            "rel": 0,
            "modestbranding": 1,
            "fs": 0,  // Disable YouTube's native fullscreen (we handle it ourselves)
            "origin": "https://www.youtube.com"
        ]

        // Override with custom player vars
        for (key, value) in playerVars {
            vars[key] = value
        }

        // Convert to JSON string
        let playerVarsJSON = vars.map { "\"\($0.key)\": \($0.value)" }.joined(separator: ", ")

        return """
        <!DOCTYPE html>
        <html>
        <head>
            <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
            <style>
                * { margin: 0; padding: 0; box-sizing: border-box; }
                html, body { width: 100%; height: 100%; background: #000; overflow: hidden; }
                #player { position: absolute; top: 0; left: 0; width: 100%; height: 100%; }
            </style>
        </head>
        <body>
            <div id="player"></div>
            <script src="https://www.youtube.com/iframe_api"></script>
            <script>
                var player;
                var progressInterval;

                function onYouTubeIframeAPIReady() {
                    player = new YT.Player('player', {
                        videoId: '\(videoId)',
                        playerVars: { \(playerVarsJSON) },
                        events: {
                            'onReady': onPlayerReady,
                            'onStateChange': onPlayerStateChange,
                            'onPlaybackQualityChange': onPlaybackQualityChange,
                            'onError': onPlayerError
                        }
                    });
                }

                function onPlayerReady(event) {
                    sendMessage('ready', {});
                    startProgressTracking();
                }

                function onPlayerStateChange(event) {
                    sendMessage('stateChange', { state: event.data });
                }

                function onPlaybackQualityChange(event) {
                    sendMessage('qualityChange', { quality: event.data });
                }

                function onPlayerError(event) {
                    sendMessage('error', { code: event.data });
                }

                function startProgressTracking() {
                    progressInterval = setInterval(function() {
                        if (player && player.getCurrentTime) {
                            var time = player.getCurrentTime();
                            sendMessage('progress', { time: time });
                        }
                    }, 500);
                }

                function sendMessage(event, data) {
                    var message = { event: event, data: data };
                    window.webkit.messageHandlers.youtubePlayer.postMessage(JSON.stringify(message));
                }
            </script>
        </body>
        </html>
        """
    }

    // MARK: - Cleanup

    deinit {
        webView?.configuration.userContentController.removeScriptMessageHandler(forName: "youtubePlayer")
    }
}

// MARK: - WKNavigationDelegate

extension YouTubePlayerView: WKNavigationDelegate {

    /// Intercept navigation to prevent redirects to YouTube app or Safari
    public func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {

        guard let url = navigationAction.request.url else {
            decisionHandler(.allow)
            return
        }

        let urlString = url.absoluteString

        // Allow YouTube iframe API and player resources
        if urlString.contains("youtube.com/iframe_api") ||
           urlString.contains("youtube.com/embed") ||
           urlString.contains("youtube.com/s/player") ||
           urlString.contains("ytimg.com") ||
           urlString.contains("googlevideo.com") ||
           urlString.hasPrefix("data:") ||
           urlString.hasPrefix("about:") {
            decisionHandler(.allow)
            return
        }

        // Block navigation to external YouTube pages (prevents app/Safari redirects)
        if urlString.contains("youtube.com/watch") ||
           urlString.contains("youtu.be/") ||
           navigationAction.navigationType == .linkActivated {
            decisionHandler(.cancel)
            return
        }

        // Allow initial HTML load and other internal navigation
        if navigationAction.navigationType == .other {
            decisionHandler(.allow)
            return
        }

        decisionHandler(.cancel)
    }
}

// MARK: - WKScriptMessageHandler

extension YouTubePlayerView: WKScriptMessageHandler {

    public func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard let body = message.body as? String,
              let data = body.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let event = json["event"] as? String,
              let eventData = json["data"] as? [String: Any] else {
            return
        }

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            switch event {
            case "ready":
                self.isReady = true
                self.delegate?.playerReady(self)

            case "stateChange":
                if let stateRaw = eventData["state"] as? Int,
                   let state = YouTubePlayerState(rawValue: stateRaw) {
                    self.delegate?.player(self, didChangeStateTo: state)
                }

            case "qualityChange":
                if let qualityRaw = eventData["quality"] as? String,
                   let quality = YouTubePlaybackQuality(rawValue: qualityRaw) {
                    self.delegate?.player(self, didChangeQualityTo: quality)
                }

            case "error":
                if let code = eventData["code"] as? Int {
                    let error = YouTubePlayerError(rawValue: code) ?? .unknown
                    self.delegate?.player(self, didReceiveError: error)
                }

            case "progress":
                if let time = eventData["time"] as? Double {
                    self.delegate?.player(self, didPlayTime: Float(time))
                }

            default:
                break
            }
        }
    }
}

// MARK: - LeakAvoider (prevents retain cycle with WKWebView)

private class LeakAvoider: NSObject, WKScriptMessageHandler {
    weak var delegate: WKScriptMessageHandler?

    init(delegate: WKScriptMessageHandler) {
        self.delegate = delegate
        super.init()
    }

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        delegate?.userContentController(userContentController, didReceive: message)
    }
}

// MARK: - Fullscreen View Controller

/// A simple fullscreen container for the YouTube player
public class YouTubeFullscreenViewController: UIViewController {

    private weak var playerView: YouTubePlayerView?
    private var originalSuperview: UIView?
    private var originalFrame: CGRect = .zero

    init(playerView: YouTubePlayerView) {
        self.playerView = playerView
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    public override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .black

        guard let playerView = playerView else { return }

        // Store original position
        originalSuperview = playerView.superview
        originalFrame = playerView.frame

        // Move player to fullscreen
        playerView.removeFromSuperview()
        playerView.frame = view.bounds
        playerView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        view.addSubview(playerView)

        // Add close button
        let closeButton = UIButton(type: .system)
        closeButton.setTitle("✕", for: .normal)
        closeButton.titleLabel?.font = .systemFont(ofSize: 24, weight: .bold)
        closeButton.setTitleColor(.white, for: .normal)
        closeButton.addTarget(self, action: #selector(closeTapped), for: .touchUpInside)
        closeButton.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(closeButton)

        NSLayoutConstraint.activate([
            closeButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 16),
            closeButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
            closeButton.widthAnchor.constraint(equalToConstant: 44),
            closeButton.heightAnchor.constraint(equalToConstant: 44)
        ])
    }

    public override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return .all
    }

    public override var prefersStatusBarHidden: Bool {
        return true
    }

    @objc private func closeTapped() {
        guard let playerView = playerView else {
            dismiss(animated: true)
            return
        }

        // Restore player to original position
        playerView.removeFromSuperview()
        playerView.frame = originalFrame
        originalSuperview?.addSubview(playerView)

        dismiss(animated: true)
    }
}

File: YouTubePlayerAndroid/app/src/main/java/com/demo/youtubeplayer/YouTubePlayerView.kt
package com.demo.youtubeplayer

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.webkit.*
import android.widget.FrameLayout

/**
 * YouTubePlayerView - A protected YouTube player for Android
 *
 * Features:
 * - No redirects to YouTube app or browser
 * - No copy link functionality
 * - No share button
 * - No text selection
 * - No context menu (long press)
 *
 * Uses WebView + YouTube IFrame API for in-app playback
 */
class YouTubePlayerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    private var webView: WebView? = null
    private var listener: YouTubePlayerListener? = null
    private var isReady = false
    private var currentVideoId: String? = null

    // Player states matching YouTube IFrame API
    enum class PlayerState(val value: Int) {
        UNSTARTED(-1),
        ENDED(0),
        PLAYING(1),
        PAUSED(2),
        BUFFERING(3),
        CUED(5)
    }

    // Playback quality options
    enum class PlaybackQuality(val value: String) {
        SMALL("small"),
        MEDIUM("medium"),
        LARGE("large"),
        HD720("hd720"),
        HD1080("hd1080"),
        HIGH_RES("highres"),
        AUTO("default")
    }

    // Error codes from YouTube IFrame API
    enum class PlayerError(val code: Int) {
        INVALID_PARAMETER(2),
        HTML5_ERROR(5),
        VIDEO_NOT_FOUND(100),
        NOT_EMBEDDABLE(101),
        NOT_EMBEDDABLE_ALT(150),
        UNKNOWN(-1)
    }

    interface YouTubePlayerListener {
        fun onReady()
        fun onStateChange(state: PlayerState)
        fun onPlaybackQualityChange(quality: PlaybackQuality)
        fun onError(error: PlayerError)
        fun onCurrentTimeChange(seconds: Float)
    }

    init {
        setupWebView()
    }

    @SuppressLint("SetJavaScriptEnabled", "ClickableViewAccessibility")
    private fun setupWebView() {
        webView = WebView(context).apply {
            layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
            setBackgroundColor(android.graphics.Color.BLACK)

            settings.apply {
                javaScriptEnabled = true
                mediaPlaybackRequiresUserGesture = false
                domStorageEnabled = true
                loadWithOverviewMode = true
                useWideViewPort = true
                cacheMode = WebSettings.LOAD_DEFAULT

                // Disable zoom
                setSupportZoom(false)
                builtInZoomControls = false
                displayZoomControls = false
            }

            // Disable long press context menu
            setOnLongClickListener { true }
            isLongClickable = false
            isHapticFeedbackEnabled = false

            // Block text selection
            setOnTouchListener { v, event ->
                if (event.action == MotionEvent.ACTION_DOWN) {
                    v.requestFocus()
                }
                false
            }

            webViewClient = object : WebViewClient() {
                override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
                    val url = request?.url?.toString() ?: return false

                    // Allow YouTube embed and player resources
                    if (url.contains("youtube.com") ||
                        url.contains("youtube-nocookie.com") ||
                        url.contains("ytimg.com") ||
                        url.contains("googlevideo.com") ||
                        url.contains("ggpht.com") ||
                        url.contains("gstatic.com") ||
                        url.contains("google.com") ||
                        url.contains("googleads") ||
                        url.contains("doubleclick") ||
                        url.startsWith("data:") ||
                        url.startsWith("about:") ||
                        url.startsWith("blob:")) {
                        // Block only watch/share URLs
                        if (url.contains("youtube.com/watch") || url.contains("youtu.be/")) {
                            return true
                        }
                        return false
                    }

                    return true // Block all other external navigation
                }

                override fun onPageFinished(view: WebView?, url: String?) {
                    super.onPageFinished(view, url)
                    // Inject CSS to disable selection after page load
                    view?.evaluateJavascript("""
                        (function() {
                            var style = document.createElement('style');
                            style.innerHTML = '* { -webkit-touch-callout: none !important; -webkit-user-select: none !important; user-select: none !important; }';
                            document.head.appendChild(style);
                        })();
                    """.trimIndent(), null)
                }
            }

            webChromeClient = object : WebChromeClient() {
                // Handle fullscreen if needed
            }

            addJavascriptInterface(YouTubeJSInterface(), "AndroidPlayer")
        }

        addView(webView)
    }

    /**
     * Load a YouTube video by ID
     */
    fun loadVideo(videoId: String) {
        currentVideoId = videoId
        isReady = false

        // Generate and load HTML with YouTube IFrame API
        val html = generateHTML(videoId)

        // Encode HTML and load as data URL - works better with YouTube IFrame API
        val encodedHtml = android.util.Base64.encodeToString(
            html.toByteArray(Charsets.UTF_8),
            android.util.Base64.NO_PADDING
        )
        webView?.loadData(encodedHtml, "text/html; charset=utf-8", "base64")
    }

    /**
     * Load a YouTube video by URL
     */
    fun loadVideoUrl(url: String) {
        extractVideoId(url)?.let { videoId ->
            loadVideo(videoId)
        }
    }

    /**
     * Play the video
     */
    fun play() {
        if (isReady) {
            webView?.evaluateJavascript("player.playVideo();", null)
        }
    }

    /**
     * Pause the video
     */
    fun pause() {
        if (isReady) {
            webView?.evaluateJavascript("player.pauseVideo();", null)
        }
    }

    /**
     * Stop the video
     */
    fun stop() {
        if (isReady) {
            webView?.evaluateJavascript("player.stopVideo();", null)
        }
    }

    /**
     * Seek to a specific time in seconds
     */
    fun seekTo(seconds: Float, allowSeekAhead: Boolean = true) {
        if (isReady) {
            webView?.evaluateJavascript("player.seekTo($seconds, $allowSeekAhead);", null)
        }
    }

    /**
     * Mute the video
     */
    fun mute() {
        if (isReady) {
            webView?.evaluateJavascript("player.mute();", null)
        }
    }

    /**
     * Unmute the video
     */
    fun unMute() {
        if (isReady) {
            webView?.evaluateJavascript("player.unMute();", null)
        }
    }

    /**
     * Set volume (0-100)
     */
    fun setVolume(volume: Int) {
        if (isReady) {
            val clampedVolume = volume.coerceIn(0, 100)
            webView?.evaluateJavascript("player.setVolume($clampedVolume);", null)
        }
    }

    /**
     * Set playback rate (0.25, 0.5, 1, 1.5, 2)
     */
    fun setPlaybackRate(rate: Float) {
        if (isReady) {
            webView?.evaluateJavascript("player.setPlaybackRate($rate);", null)
        }
    }

    /**
     * Set playback quality
     */
    fun setPlaybackQuality(quality: PlaybackQuality) {
        if (isReady) {
            webView?.evaluateJavascript("player.setPlaybackQuality('${quality.value}');", null)
        }
    }

    /**
     * Load a new video while player is ready
     */
    fun cueVideoById(videoId: String) {
        currentVideoId = videoId
        if (isReady) {
            webView?.evaluateJavascript("player.cueVideoById('$videoId');", null)
        }
    }

    /**
     * Load and play a new video immediately
     */
    fun loadVideoById(videoId: String) {
        currentVideoId = videoId
        if (isReady) {
            webView?.evaluateJavascript("player.loadVideoById('$videoId');", null)
        }
    }

    /**
     * Set listener for player events
     */
    fun setYouTubePlayerListener(listener: YouTubePlayerListener) {
        this.listener = listener
    }

    /**
     * Clean up resources
     */
    fun release() {
        webView?.apply {
            stopLoading()
            clearHistory()
            removeAllViews()
            destroy()
        }
        webView = null
    }

    private fun extractVideoId(url: String): String? {
        // Handle youtu.be short URLs
        if (url.contains("youtu.be/")) {
            return url.substringAfter("youtu.be/").substringBefore("?")
        }

        // Handle youtube.com/watch URLs
        if (url.contains("youtube.com/watch")) {
            val uri = android.net.Uri.parse(url)
            return uri.getQueryParameter("v")
        }

        // Handle youtube.com/embed URLs
        if (url.contains("/embed/")) {
            return url.substringAfter("/embed/").substringBefore("?")
        }

        return null
    }

    private fun generateHTML(videoId: String): String {
        return """
            <!DOCTYPE html>
            <html>
            <head>
                <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
                <style>
                    * {
                        margin: 0;
                        padding: 0;
                        box-sizing: border-box;
                        -webkit-touch-callout: none !important;
                        -webkit-user-select: none !important;
                        user-select: none !important;
                    }
                    html, body {
                        width: 100%;
                        height: 100%;
                        background: #000;
                        overflow: hidden;
                    }
                    #player {
                        position: absolute;
                        top: 0;
                        left: 0;
                        width: 100%;
                        height: 100%;
                    }
                    /* HIDE ALL YOUTUBE LOGOS AND BRANDING */
                    .ytp-watermark, .ytp-watermark-container {
                        display: none !important;
                        visibility: hidden !important;
                        opacity: 0 !important;
                        pointer-events: none !important;
                    }
                    .ytp-youtube-button, .ytp-button[aria-label*="YouTube"],
                    a[href*="youtube.com"], .ytp-title-link {
                        display: none !important;
                        visibility: hidden !important;
                    }
                    .ytp-share-button { display: none !important; }
                    .ytp-watch-later-button { display: none !important; }
                    .ytp-pause-overlay, .ytp-pause-overlay-container { display: none !important; }
                    .ytp-endscreen-content, .ytp-endscreen-previous, .ytp-endscreen-next,
                    .ytp-ce-element, .ytp-ce-covering-overlay { display: none !important; }
                    .ytp-cards-button, .ytp-cards-teaser { display: none !important; }
                    .ytp-title, .ytp-title-text, .ytp-title-channel,
                    .ytp-title-channel-logo, .ytp-chrome-top { display: none !important; }
                    .ytp-impression-link, .ytp-show-cards-title { display: none !important; }
                    .ytp-overflow-button { display: none !important; }

                    /* Overlay to block logo clicks */
                    #logo-blocker {
                        position: absolute;
                        bottom: 0;
                        right: 0;
                        width: 120px;
                        height: 50px;
                        z-index: 9998;
                        background: transparent;
                        pointer-events: auto;
                    }
                </style>
            </head>
            <body>
                <div id="player"></div>
                <div id="logo-blocker"></div>
                <script src="https://www.youtube.com/iframe_api"></script>
                <script>
                    var player;
                    var progressInterval;

                    // Disable context menu
                    document.addEventListener('contextmenu', function(e) {
                        e.preventDefault();
                        e.stopPropagation();
                        return false;
                    }, true);

                    // Disable selection
                    document.addEventListener('selectstart', function(e) {
                        e.preventDefault();
                        return false;
                    });

                    // Disable copy
                    document.addEventListener('copy', function(e) {
                        e.preventDefault();
                        return false;
                    });

                    function onYouTubeIframeAPIReady() {
                        player = new YT.Player('player', {
                            videoId: '$videoId',
                            playerVars: {
                                'playsinline': 1,
                                'controls': 1,
                                'showinfo': 0,
                                'rel': 0,
                                'modestbranding': 1,
                                'fs': 0,
                                'disablekb': 0,
                                'enablejsapi': 1
                            },
                            events: {
                                'onReady': onPlayerReady,
                                'onStateChange': onPlayerStateChange,
                                'onPlaybackQualityChange': onPlaybackQualityChange,
                                'onError': onPlayerError
                            }
                        });
                    }

                    function onPlayerReady(event) {
                        AndroidPlayer.onReady();
                        startProgressTracking();
                    }

                    function onPlayerStateChange(event) {
                        AndroidPlayer.onStateChange(event.data);
                    }

                    function onPlaybackQualityChange(event) {
                        AndroidPlayer.onPlaybackQualityChange(event.data);
                    }

                    function onPlayerError(event) {
                        AndroidPlayer.onError(event.data);
                    }

                    function startProgressTracking() {
                        progressInterval = setInterval(function() {
                            if (player && player.getCurrentTime) {
                                var time = player.getCurrentTime();
                                AndroidPlayer.onCurrentTime(time);
                            }
                        }, 500);
                    }
                </script>
            </body>
            </html>
        """.trimIndent()
    }

    /**
     * JavaScript interface for communication from WebView to Android
     */
    private inner class YouTubeJSInterface {

        @JavascriptInterface
        fun onReady() {
            post {
                isReady = true
                listener?.onReady()
            }
        }

        @JavascriptInterface
        fun onStateChange(state: Int) {
            post {
                val playerState = PlayerState.values().find { it.value == state } ?: PlayerState.UNSTARTED
                listener?.onStateChange(playerState)
            }
        }

        @JavascriptInterface
        fun onPlaybackQualityChange(quality: String) {
            post {
                val playbackQuality = PlaybackQuality.values().find { it.value == quality } ?: PlaybackQuality.AUTO
                listener?.onPlaybackQualityChange(playbackQuality)
            }
        }

        @JavascriptInterface
        fun onError(errorCode: Int) {
            post {
                val error = PlayerError.values().find { it.code == errorCode } ?: PlayerError.UNKNOWN
                listener?.onError(error)
            }
        }

        @JavascriptInterface
        fun onCurrentTime(seconds: Float) {
            post {
                listener?.onCurrentTimeChange(seconds)
            }
        }
    }
}
